### ISSUE TITLE
BUG: IndexError: list index out of range

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/axondeepseg#34]
data: https://osf.io/28k4m/

~~~
(ads_venv) julien-macbook:~/Desktop/im $ axondeepseg -t SEM -p data.png -v 2
Loading acquisitions ...
Rescaling acquisitions to the target resolution ...
Graph construction ...
Beginning inference ...
Traceback (most recent call last):
  File "/Users/julien/code/ads_venv/bin/axondeepseg", line 11, in <module>
    sys.exit(main())
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/segment.py", line 225, in main
    verbosity_level=verbosity_level)
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/segment.py", line 57, in segment_image
    prediction_proba_activate=False, write_mode=True)
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/apply_model.py", line 243, in axon_segmentation
    gpu_per=gpu_per, verbosity_level=verbosity_level)
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/apply_model.py", line 165, in apply_convnet
    verbose_mode=0)
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/apply_model.py", line 360, in process_segmented_patches
    patch_size = predictions_list[0].shape[0]
IndexError: list index out of range
~~~

Another bug:
~~~
axondeepseg -t TEM -p data.png -v 2
Loading acquisitions ...
Rescaling acquisitions to the target resolution ...
Graph construction ...
Beginning inference ...
Traceback (most recent call last):
  File "/Users/julien/code/ads_venv/bin/axondeepseg", line 11, in <module>
    sys.exit(main())
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/segment.py", line 225, in main
    verbosity_level=verbosity_level)
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/segment.py", line 57, in segment_image
    prediction_proba_activate=False, write_mode=True)
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/apply_model.py", line 243, in axon_segmentation
    gpu_per=gpu_per, verbosity_level=verbosity_level)
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/apply_model.py", line 107, in apply_convnet
    prediction_proba_activate=prediction_proba_activate)
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/AxonDeepSeg/apply_model.py", line 433, in perform_batch_inference
    tf_input: batch_x})  # here we get the predictions under prob format (float, between 0 and 1, shape = (bs, size_image*size_image, n_classes).
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 889, in run
    run_metadata_ptr)
  File "/Users/julien/code/ads_venv/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1096, in _run
    % (np_val.shape, subfeed_t.name, str(subfeed_t.get_shape())))
ValueError: Cannot feed value of shape (1, 122, 120) for Tensor u'Placeholder:0', which has shape '(?, 512, 512)'
~~~

**Comment by alzaia** [comment]
Coming back to this issue, it comes from the patch size input (512x512) after resampling to the target pixel size. I added a warning in the doc last December (https://neuropoly.github.io/axondeepseg/documentation.html#using-axondeepseg). However, we need to find a better solution. 

I was discussing about this with @perone and it's a complex issue. The easy solution would be to throw an exception when this happens. The best solution would be to add padding to the sample, however this is tricky because it would require revalidation of everything (what kind of padding to use, what will happen to the segmentation at borders between image and padding as the padding will change the contextual information that the network is given to perform the task,...).

**Comment by jcohenadad** [comment]
ok, so maybe just throwing an exception would be more elegant, and telling the user what to do if that happens.

**Comment by mathieuboudreau** [comment]
I just revisited this, and I think this specific issue is not a high priority anymore nor what I would call a bug anymore. My suggestion, as I'll argue below, would be lower this issue priority to low and reference it in issue #29 to be resolved there.

What we need to do, as hinted in a few comments above, is deal with this in the exception handling issue #29  (which is barebones and/or non-existant at the moment). Since SEM and TEM models have a fixed patch size (256 and/or 512 depending on the settings), and ADS always rescales the images to fixed resolutions (0.1 for SEM data, 0.01 for TEM), what we need to do is calculate what the dimensions of the input image(s) will after rescaling it by using the CLI-given image resolution, and if the calculated rescaling dimensions are smaller than the patch size, throw an exception message to the user. 

Both errors that @jcohenadad encountered are due to different degrees/order of magnitudes of rescaling causing too small images for the expected patch generation. For the sample image above, if I call ADS using the actual image resolution flag (which you weren't in your commands above, maybe because they couldn't be set back then?), then the segmentation works quite well with no issues:

`axondeepseg -t TEM -i data.tif -v 2 -s 0.004` 

<details>

![data_seg-axonmyelin](https://user-images.githubusercontent.com/1421029/43554142-2ed6f804-95c9-11e8-9fe4-19541289b52e.png)

</details><p>

Whereas, if I set the image resolution too small (e.g. since the image dimension is 1888 x 2304, giving it a pixel size flag of 0.001 will rescale it to 189 x 230), it will cause the following error to occur:

`axondeepseg -t TEM -i data.tif -v 2 -s 0.001`

causes:

```
...
ValueError: Cannot feed value of shape (1, 230, 189) for Tensor u'Placeholder:0', which has shape '(?, 512, 512)'
```

If I make the resolution one order smaller than that (-s 0.0001, rescaling it to 19x23), then the other error occurs:

`axondeepseg -t TEM -i data.tif -v 2 -s 0.0001`

causes:

```
...
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/apply_model.py", line 397, in process_segmented_patches
    patch_size = predictions_list[0].shape[0]
IndexError: list index out of range
```

Both of which can be avoided by the exception handling I suggested above (immediately after the CLI arguments are parsed). This should be handled in issue #29, and maybe we could increase the priority level of that one in exchange of lowering this one.
